Ensure the Transaction Manager node name is less than or equal to 28 bytes

marcosgopen opened the pull request:
We need to include the node name in the fixed size Xid data structure(limited to 28 bytes) so that the TM can determine which XA resource branches it is responsible for.

related to PR: https://github.com/quarkusio/quarkus/pull/36752

please note: this is fixing the issue truncating the byte array which means that it impacts the collision avoidance. This is why I added a stress test to 'test' the collision. I think that the risk of collision is still very low but a preferable solution may be to have a byte array type 'nodeName' so that the collision resistance exclusively relies on the hash algorithm.  

- Resolves: #40256

marcosgopen commented:
FYI @turing85 and @mmusgrov

mmusgrov commented:
I added a [comment to the issue](https://github.com/quarkusio/quarkus/issues/40256#issuecomment-2109810626) indicating that the spring boot port should just avoid using the Unicode character set and use ISO-LATIN-1 instead. This would mean the original code submitted by @turing85 would stand.

mmusgrov reviewed (changes requested):
I am declining this PR because the node name is less than or equal to 28 bytes when a fixed-length character encoding is used to convert it to a byte array.

marcosgopen commented:
Closed as 'not a bug'

turing85 commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaRecorder.java:
```diff
@@ -57,9 +58,19 @@ private static void shortenNodeName(TransactionManagerConfiguration transactions
+        //and the conversion is reversible
+        byte[] base64Result = Base64.getEncoder().encode(hashedByteArray);
+        //truncate the array
+        byte[] slice = Arrays.copyOfRange(base64Result, 0, 27);
```
Shouldn't the `27` be a `28` since it is exclusive?

geoand commented:
This is going to need a better title

yrodiere reviewed (commented):
LGTM, though I'll wait for @mmusgrov's approval before actually approving -- he knows the implications in Narayana, I really don't.

marcosgopen commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/NarayanaJtaRecorder.java:
```diff
@@ -57,9 +58,19 @@ private static void shortenNodeName(TransactionManagerConfiguration transactions
-        final byte[] nodeNameAsBytes = originalNodeName.getBytes();
+        final byte[] nodeNameAsBytes = originalNodeName.getBytes(StandardCharsets.UTF_8);
         MessageDigest messageDigest224 = MessageDigest.getInstance(HASH_ALGORITHM_FOR_SHORTENING);
-        transactions.nodeName = new String(messageDigest224.digest(nodeNameAsBytes), StandardCharsets.UTF_8);
```
Thanks @yrodiere for your analysis. I would like to clarify some points (I hope we all agree about them):
- only Strings longer then 28 bytes are shorten by the following process.
- the original string has an arbitrary length ( >28 bytes), then it is hashed and it's output is 28 bytes (SHA-224). This is the only hash used in this process.
- Base64 encoding is done after the byte array is hashed.
- the base64 encoded byte array has an arbitrary length and so it is truncate so that it is a 28 bytes array 
- Using the Base64 encoding for the byte array the resulting Strings are printable, are utf-8 valid and can be easily encoded to UTF-8 without changing the length

graben replied:
Shouldn't this also be documented in the JavaDoc of https://github.com/quarkusio/quarkus/blob/2d4b470b10091632cee1bea4a44380a85bc71baa/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java#L35 ?

graben replied:
Btw. I heard about different numbers that are actually known limits.  28 for Narayana itself and 23 for usage under Kubernetes. Would it make sense to make the actual algorithm more configurable. E.g. a number property with valid range from 1 to 28 and 28 as default. Maybe even replace `TransactionManagerConfiguration.shortenNodeNameIfNecessary` and let the default of the new property be 0 to signal no shortening at all.

marcosgopen replied:
I will update the JavaDoc, thanks @graben . Concerning having a parameter to configure the length of the string I think it might be risky because the hash collision would be directly impacted when to string length is changed. So  we would need to make sure that the users know exactly about the impact of that configuration's change. However we need to remember that only string with byte array's length > 28 are shorten, other strings are not. So when the users have a strict limitation about the string length (e.g. for Kubernetes 23 as you said)  they can pass 23 bytes string and that string will not be changed.

marcosgopen replied:
Good point @graben. Yes, it does. But I would say that the collision avoidance with 28 bytes is still acceptable (see the stress test) IMO, but it probably would not with a random number of bytes between 1 to 27. So again I am OK to make that number configurable but I would recommend anyway to shorten the hash with the longest string allowed (which is 28).

graben replied:
IMHO, the question is if 28 is the maximum allowed by Narayana, why offer an optional  property to shorten values that are obviously wrong at all? Shouldn't Quarkus disallow values larger than 28 and force users to shorten themselves? No potential problems with collisions anymore! I know it's a bit of a radical solution, but it feels odd that users can define invalid properties and the software should solve it with a workaround with potential drawbacks. Just my 2 cents.

marcosgopen replied:
Thanks @graben for your comment here. I am not sure I am following you here. You first were talking about adding an extra parameter for the length of the node name, now you are talking about removing this 'Generate right-length node name' feature (the one discussed in the issue https://github.com/quarkusio/quarkus/issues/30491 ). I might be wrong here but this PR's goal is to patch the existing implementation, not to change it. Could I maybe ask you to discuss this in a different issue or on Zulip where it is easier to get the community's feedback about this topic?
Getting back to your comment I would like to underline that the default behaviour is the one that you described: in fact the 'shortenNodeNameIfNecessary' variable is set to false by default (see https://github.com/quarkusio/quarkus/blob/main/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java#L34), the original reasons to implement this feature are discussed in the issue https://github.com/quarkusio/quarkus/issues/30491.

graben replied:
Keep it as is, it just came into my mind, while following the process involving from the very first beginning. ;-)

mmusgrov replied:
@graben and @marcosgopen I 'm not asking for changes here, but for completeness/clarification the 28 byte limit is needed only for XA, I think, and the 23 byte limit, so far, is only for kubernetes. But Narayana supports a number of non-XA transaction models and for most of those the size limit is 64 bytes. The 23 byte limit was introduced because the Wildfly k8s integration relies on Stateful sets and a Stateful id is 63 bytes and k8s sets the final 5 characters randomly (so we need 28 bytes minus 5 bytes for the maximum length of the nodeIdentifier - I forget why that forces it to be reduced, I'd have to have another think about why the original engineers noticed that as as being an issue).

marcosgopen replied:
This is a good question @mmusgrov . As I wrote to Benjamin this PR patches https://github.com/quarkusio/quarkus/pull/36752 . If we want to be able to set the node name size configurable I would suggest to open a new PR or rethink this PR. Keep in mind that allowing the node name size to be configurable might increase the risk of name collision.

mmusgrov replied:
@marcosgopen I wasn't asking a question, I was only "tying up loose ends" and providing clarity on why we enforce different size limits depending upon the particular transaction model and/or environment - i.e. it was a comment designed to contribute to the knowledge base, I wasn't asking for changes.

graben replied:
Hi @mmusgrov, just a small question according to the 28 `byte` limit for XA. As far as I understand the code of the [`XID`](https://github.com/jbosstm/narayana/blob/0be17bd96db646b47b93b69ee6855eed8a39aeda/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/xa/XID.java#L28) implementation, it needs `gtrid` to be less than 64. In [`XATxConverter`](https://github.com/jbosstm/narayana/blob/main/ArjunaJTA/jta/classes/com/arjuna/ats/jta/xa/XATxConverter.java) `gtrid` is computed by concatenating `UID` as `byte` and `nodeIdentifier` as `bytes` both limited to 28 `byte`. That's a maximum of 56 < 64. Why does Narayana reserve a free space of 8 `byte`s length? Maybe I do miss some other point in the code.

graben replied:
Hi @mmusgrov, you might missed my question. :-)

marcosgopen replied:
Hi @graben , thanks for your question. The answer is that "in reality this must be less that 28 to fit in the bqual with the EIS name and Uid." see https://issues.redhat.com/browse/JBTM-1651

marcosgopen replied:
@yrodiere Do you have more questions about this or you are OK to resolve this thread?

yrodiere replied:
I'm fine with resolving this thread, I was merely pointing out that it was important to merge the solution as-is for more reasons than just https://github.com/quarkusio/quarkus/issues/40256.
The patch was already good as far as I'm concerned :)

gastaldi commented on extensions/narayana-jta/runtime/src/test/java/io/quarkus/narayana/jta/runtime/NarayanaJtaRecorderTest.java:
```diff
@@ -0,0 +1,60 @@
+            transactions.nodeName = originalNodeName;
+            r.setNodeName(transactions);
+            int numberOfBytes = transactions.nodeName.getBytes(StandardCharsets.UTF_8).length;
+            if (numberOfBytes > 28) {
+                fail("node name bytes still exceeded 28 bytes limit, number of bytes is " + numberOfBytes);
+            }
```
```suggestion
            assertFalse(numberOfBytes > 28, "node name bytes still exceeded 28 bytes limit, number of bytes is " + numberOfBytes);
```

marcosgopen replied:
Thanks @gastaldi for your suggestion , I've updated the test.

gastaldi commented on extensions/narayana-jta/runtime/src/test/java/io/quarkus/narayana/jta/runtime/NarayanaJtaRecorderTest.java:
```diff
@@ -0,0 +1,60 @@
+            transactions.nodeName = new String(data);
+            r.setNodeName(transactions);
+            int numberOfBytes = transactions.nodeName.getBytes(StandardCharsets.UTF_8).length;
+            if (numberOfBytes > 28) {
+                fail("node name bytes still exceeded 28 bytes limit, number of bytes is " + numberOfBytes);
+            }
```
```suggestion
            assertFalse(numberOfBytes > 28, "node name bytes still exceeded 28 bytes limit, number of bytes is " + numberOfBytes);
```

marcosgopen replied:
Thanks @gastaldi for your suggestion , I've updated the test.

marcosgopen commented on extensions/narayana-jta/runtime/src/test/java/io/quarkus/narayana/jta/runtime/NarayanaJtaRecorderTest.java:
```diff
@@ -0,0 +1,60 @@
+            r.setNodeName(transactions);
+            if (set.contains(transactions.nodeName)) {
+                fail("Collision of IDs for " + transactions.nodeName);
+            }
```
Thanks @gastaldi for your suggestion , I've updated the test.

mmusgrov commented:
@graben We created and merged a narayana PR for https://issues.redhat.com/browse/JBTM-3883 (Add a setNodeIdentifier config method that accepts a byte array) which should render this PR superfluous but it would need to be replaced with a change that calls the new [setNodeIdentifier config method that accepts a byte array](https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/common/CoreEnvironmentBean.java#L136).

graben commented:
Hi @mmusgrov, thanks for the info. Does exist a timeline for the new narayana release containing this patch?

mmusgrov commented:
All being well, we should be releasing a version that contains that change this week.

FYI The test that validates the new setting is https://github.com/jbosstm/narayana/blob/main/ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/xa/NodeIdentifierUnitTest.java#L38

mmusgrov commented:
I'd recommend leaving this PR open until the new patch has been tested against Quarkus.

geoand commented:
👍🏼

marcosgopen commented:
Could I ask you to retest this PR now that https://github.com/quarkusio/quarkus/pull/41539 has been merged? Thanks @geoand

geoand commented:
I'm just facilitating things here 😀

marcosgopen commented:
> > Could I ask you to retest this PR now that #41539 has been merged? Thanks @geoand
> 
> @marcosgopen don't we need to change the call to set the node name to use the [new method](https://github.com/jbosstm/narayana/blob/7.0.2.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/common/CoreEnvironmentBean.java#L136). If we do this then we won't be needing this PR, ie there isn't any benefit in retesting this PR. I suggested leaving the PR open only until we have tested that quarkus works with the new method.

I agree with you @mmusgrov , let's keep this PR as is until we can test the [new method](https://github.com/jbosstm/narayana/blob/7.0.2.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/common/CoreEnvironmentBean.java#L136), and then I can update this PR to use that method.

mmusgrov commented:
> I agree with you @mmusgrov , let's keep this PR as is until we can test the [new method](https://github.com/jbosstm/narayana/blob/7.0.2.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/common/CoreEnvironmentBean.java#L136), and then I can update this PR to use that method.

The updated PR will be the test for the new method, ie we have already tested the new method in Narayana but now we need to test that it works for quarkus as well.

mmusgrov commented:
@graben and @geoand FYI the Narayana fix for setting the node identifier bytes directly suffers from the same problem that this PR is fixing. This means that we can go ahead with Marco's fix (to use a base64 encoding even after shortening with SHA-224). I will re-review this PR and independently I will update Narayana to use the same base64 encoding - ie this PR does not need to be blocked, thanks @marcosgopen for supplying the fix.,

graben commented:
Hi @mmusgrov, do I understand right that shortening a base64 sha224 hashed string should be used, even if setting byte[] in Narayana directly?

mmusgrov commented:
@graben Setting byte[] in Narayana directly is not right because Narayana needs the same fix that this PR is proposing - so my proposal is to not block this PR. I will re-review this PR with the aim of approving it.

graben commented:
> @graben Setting byte[] in Narayana directly is not right because Narayana needs the same fix that this PR is proposing - so my proposal is to not block this PR. I will re-review this PR with the aim of approving it.

Okay, I already adapted the new method setNodeIdentifier(byte[]) in Narayana for narayana-spring-boot. Should I wait until a new Narayana release (7.0.3?) is released, or should I do a base64 encoding stripped to 28 characters in the meantime?

mmusgrov commented:
> Okay, I already adapted the new method setNodeIdentifier(byte[]) in Narayana for narayana-spring-boot. Should I wait until a new Narayana release (7.0.3?) is released, or should I do a base64 encoding stripped to 28 characters in the meantime?

We will fix it in the next minor release, since it includes a new feature, which will be 7.1.0.Final.

Narayana uses the bytes set by the new method to set the nodeIdentifier string using a UTF_8 encoding [1] and if the resulting string is greater than 64 bytes (XA support requires it to be < 28 but all other transaction models only require it to be < 64), which only happens with a probability of about 1 in a 100 depending upon how you generated the byte array, then an exception will be thrown so you will know that it has failed.

So it depends upon how awkward that kind of failure rate will be for your users. Personally I would do the base64 encoding striping yourself while you wait for 7.1.0.Final which I would hope should be asap, but if you code it in such a way that nodeIdentifier.getBytes(StandardCharsets.UTF_8) < 28 [3] then the standard CoreEnvironmentBean.setNodeIdentifier(String nodeIdentifier) method would be sufficient.

FYI Marco discovered the narayana issue by running the test [2] in a loop of 1000.

[1] https://github.com/jbosstm/narayana/blob/7.0.2.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/common/CoreEnvironmentBean.java#L149-L153
[2] https://github.com/jbosstm/narayana/blob/7.0.2.Final/ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/xa/NodeIdentifierUnitTest.java#L30-L34
[3] https://github.com/jbosstm/narayana/blob/7.0.2.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TxControl.java#L23

mmusgrov commented:
@graben I would like to further clarify my last comment - the Narayana fix for setting the node identifier bytes directly suffers from the same problem that this PR is fixing and the only way to make it work is to further shorten the byte array which is the same as what this PR is doing. Therefore there is no benefit in adding the setNodeIdentifier(byte[])  method to Narayana. We have closed [the Narayana issue tracker](https://issues.redhat.com/browse/JBTM-3883) and marked it as "Won't Do".

marcosgopen commented:
Thanks @mmusgrov , I will update the PR and also the doc (https://github.com/quarkusio/quarkus/blob/main/docs/src/main/asciidoc/transaction.adoc#configuring-transaction-node-name-identifier) so to mention the length limitation and the availability of the variable 'shortenNodeNameIfNecessary', which, when se to true, will shorten name longer than 28 bytes. In fact I couldn't find any reference about the node name length limitation in the Quarkus doc (e.g. https://github.com/quarkusio/quarkus/blob/main/docs/src/main/asciidoc/transaction.adoc#configuring-transaction-node-name-identifier ).

marcosgopen commented:
I have rebased the PR and updated it. Now the doc includes the information about the node-name limitation and shortenNodeNameIfNecessary property.
cc @mmusgrov

marcosgopen committed:
node-name needs to be truncated because the byte array is not base64

marcosgopen commented on docs/src/main/asciidoc/transaction.adoc:
```diff
@@ -307,6 +307,9 @@ The node name identifier needs to be unique per transaction manager deployment.
 The node name identifier may be configured via the property `quarkus.transaction-manager.node-name`.
+The node-name property has a length limitation, it cannot be longer than 28 bytes.
+The length limitation can possibly be solved by enabling the property 'quarkus.transaction-manager.shortenNodeNameIfNecessary' which, when enabled, will shorten name longer than 28 bytes.
+The shortening process is done by applying the 'SHA-224' algorithm, encoding to Base64 and then truncating the result.
```
Thanks @tomjenkinson , I have updated the doc.

tomjenkinson reviewed (approved):
LGTM - thank you

yrodiere commented on docs/src/main/asciidoc/transaction.adoc:
```diff
@@ -307,6 +307,14 @@ The node name identifier needs to be unique per transaction manager deployment.
 The node name identifier may be configured via the property `quarkus.transaction-manager.node-name`.
+[NOTE]
+====
+The node-name property has a length limitation, it cannot be longer than 28 bytes.
+
+The length limitation can possibly be solved by enabling the property 'quarkus.transaction-manager.shortenNodeNameIfNecessary' which, when enabled, will shorten names longer than 28 bytes.
+The shortening process is done by applying the 'SHA-224' algorithm, encoding to Base64 and then truncating the result. Therefore this may result in the node-name ending as an unpredictable duplicate string and this breaks the uniqueness that is mandatory for safe transaction usage.
+For more information about the potential for conflict see https://github.com/quarkusio/quarkus/issues/30491  
```
I would simplify the phrasing and would stress that collisions are very unlikely -- see suggestion below, but up to you.

More importantly, Quarkus automatically uses a hyphen-delimited format instead of camel case in `application.properties`, so the actual name of the property is `quarkus.transaction-manager.shorten-node-name-if-necessary`, and you need to fix that (I did below).

```suggestion
The node name cannot be longer than 28 bytes.
To automatically shorten names longer than 28 bytes, set `quarkus.transaction-manager.shorten-node-name-if-necessary` to `true`.

Shortening is implemented by hashing the node name, encoding the hash to Base64 and then truncating the result. As with all hashes, the resulting shortened node name could potentially conflict with another shortened node name, but it is https://github.com/quarkusio/quarkus/issues/30491#issuecomment-1537247764[very unlikely].
```

yrodiere commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/TransactionManagerConfiguration.java:
```diff
@@ -27,7 +27,7 @@ public final class TransactionManagerConfiguration {
      * The node name must not exceed a length of 28 bytes. If this property is set to {@code true}, and the node name exceeds 28
      * bytes, the node name is shortened by calculating the <a href="https://en.wikipedia.org/wiki/SHA-2">SHA-224</a> hash,
-     * which has a length of 28 bytes.
+     * which has a length of 28 bytes, encoded to Base64 format and then shorten to 28 bytes.
```
```suggestion
     * which has a length of 28 bytes, encoded to Base64 format and then shortened to 28 bytes.
```

yrodiere reviewed (changes requested):
Hey,

The implementation LGTM, thanks. Documentation contains a mistake, though: the name of the property.

I would also simplify phrasing in the documentation, but feel free to ignore that suggestion.

marcosgopen commented:
> Hey,
> 
> The implementation LGTM, thanks. Documentation contains a mistake, though: the name of the property.
> 
> I would also simplify phrasing in the documentation, but feel free to ignore that suggestion.

Thank you very much @yrodiere , updated and rebase.

yrodiere commented:
Tom approved from the Narayana side, and the Quarkus side LGTM: let's merge once CI passes. Thanks!

yrodiere commented:
> ## Failures
> ### ⚙️ JVM Tests - JDK 17 [#](#user-content-test-failures-job-31877859215)
> 
> ```diff
> - Failing: extensions/hibernate-reactive/deployment 
> ! Skipped: extensions/panache/hibernate-reactive-panache-common/deployment extensions/panache/hibernate-reactive-panache-kotlin/deployment extensions/panache/hibernate-reactive-panache/deployment and 21 more
> ```
> 
> #### 📦 extensions/hibernate-reactive/deployment
> 
> ✖ `Failed to execute goal io.fabric8:docker-maven-plugin:0.45.1:start (docker-start) on project quarkus-hibernate-reactive-deployment: I/O Error`

This looks like an infra issue/flake, I restarted that build.
